README.md
# yahtzee-clj

Simple Yahtzee game.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-10-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-10-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/yahtzee-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/yahtzee-clj/compare/0.1.0...0.1.1


test/yahtzee_clj/core_test.clj
(ns yahtzee-clj.core-test
  (:require [clojure.test :refer :all]
            [yahtzee-clj.core :refer :all]))

(deftest test-hand
  ;; TODO; understand the value of doing this might not always be true
  (testing "Tests length, max & min of hand."
    (is (= 5 (count (roll-hand))))
    (is (> (apply min (roll-hand)) 0))
    (is (< (apply max (roll-hand)) 7))
    )
  )

;; TODO: better docs as these are redundant?
(deftest test-score-chance
  (testing "Tests the sum of the singles"
    (is (= 15 (score-chance '(1 2 3 4 5))))
    (is (= 30 (score-chance '(6 6 6 6 6))))
    )
  )

(deftest test-score-singles
  (testing "Tests the singles"
    (is (= 3 (score-singles '(1 2 1 1 3) 1)))
    (is (= 1 (score-singles '(1 2 3 4 5) 1)))
    (is (= 5 (score-singles '(1 2 3 4 5) 5)))
    (is (= 6 (score-singles '(2 2 3 4 2) 2)))
    (is (= 6 (score-singles '(2 2 3 4 2) 2)))
    (is (= 5 (score-singles '(1 1 1 1 1) 1)))
    (is (= 0 (score-singles '(1 1 1 1 1) 6)))
    )
  )

(deftest test-score-of-a-kind
  (testing "Tests the score when choosing x of a kind"
    (is (= 3 (score-of-a-kind '(1 6 1 6 1) 3)))
    (is (= 50 (score-of-a-kind '(6 6 6 6 6) 5)))
    (is (= 50 (score-of-a-kind '(1 1 1 1 1) 5)))
    (is (= 18 (score-of-a-kind '(6 6 1 6 1) 3)))
    (is (= 0 (score-of-a-kind '(1 2 5 6 1) 3)))
    (is (= 3 (score-of-a-kind '(1 1 1 1 1) 3)))
    (is (= 4 (score-of-a-kind '(1 1 1 1 1) 4)))
    (is (= 50 (score-of-a-kind '(1 1 1 1 1) 5)))
    )
  )

(deftest test-score-full-house
  (testing
    (is (= 25 (score-full-house '(1 6 1 6 1))))
    (is (= 25 (score-full-house '(5 5 5 2 2))))
    (is (= 0 (score-full-house '(1 6 1 6 3))))
    (is (= 0 (score-full-house '(5 5 5 2 1))))
    )
  )


(deftest test-hand-to-sorted-string
  (testing
    (is (= (hand-to-sorted-string '(1 2 3 4 5)) "12345"))
    (is (= (hand-to-sorted-string '(1 1 1 1 1)) "1"))
    (is (= (hand-to-sorted-string '(2 3 4 5 5)) "2345"))
    (is (= (hand-to-sorted-string '(2 2 3 4 5)) "2345"))
    (is (= (hand-to-sorted-string '(2 3 4 5 6)) "23456"))
    (is (= (hand-to-sorted-string '(1 2 2 3 1)) "123"))
    )
  )

(deftest test-score-straight
  (testing
    (is (= 30 (score-straight '(1 2 3 4 5) 4)))
    (is (= 30 (score-straight '(5 6 1 4 3) 4)))
    (is (= 0 (score-straight '(5 6 1 4 2) 4)))
    (is (= 0 (score-straight '(5 6 1 4 1) 4)))
    (is (= 0 (score-straight '(5 6 1 4 1) 5)))
    (is (= 40 (score-straight '(5 6 3 2 4) 5)))
    (is (= 40 (score-straight '(2 1 3 5 4) 5)))
    ))

(deftest test-score-hand
  (testing
    (is (= (score-hand '(1 2 3 3 3))
            {:chance 12
             :ones 1
             :twos 2
             :threes 9
             :fours 0
             :fives 0
             :sixes 0
             :three-of-a-kind 9
             :four-of-a-kind 0
             :full-house 0
             :small-straight 0
             :large-straight 0
             :yahtzee 0
             }
           ))))

(run-all-tests)

src/yahtzee_clj/core.clj
(ns yahtzee-clj.core
  (:gen-class))
(use '[clojure.string :only (join includes?)])

(defn roll []
  "Rolls a six sided die once."
  (+ 1 (rand-int 6)))

(defn roll-hand []
  "Rolls a hand of 5 six sided dice."
  (repeatedly 5 roll))

(defn score-chance [hand]
  "Scores the chance - i.e. sum of hand."
  (reduce + hand))

(defn score-singles [hand single]
  "Scores the singles"
  (reduce + (filter #(= single %) hand))
  )

(defn score-of-a-kind [hand kind-type]
  "Scores three, four and five of a kind"
  (let [most-common (last (sort-by val (frequencies hand)))
        face (first most-common)
        count (last most-common)]
    (cond
      (< count 3) 0
      (and (>= count 3) (= kind-type 3)) (* face 3)
      (and (>= count 4) (= kind-type 4)) (* face 4)
      (and (= count 5) (= kind-type 5)) 50
      :else 0
      )
    ))

;; TODO: this is horrible!
(defn score-full-house [hand]
  "Scores three, four and five of a kind"
  (let [freqs (sort-by val (frequencies hand))
        most-common (last (last freqs))
        second-most-common (last (last (butlast freqs)))]
    (if (and
          (= most-common 3)
          (= second-most-common 2))
      25 0)))

;; TODO: use threading macro to make it easier to read?
(defn hand-to-sorted-string [hand]
  "Returns a hand into a sorted string having removed dupes"
  (join "" (sort (into [] (set (map str hand)))))
  )

(defn score-straight [hand straight_length]
  "Scores small and large straight"
  (let [valid_small '("1234" "2345" "3456")
        valid_large '("12345" "23456")]
    (cond
      (and (= straight_length 4) (some #(includes? (hand-to-sorted-string hand) %) valid_small)) 30
      (and (= straight_length 5) (some #(includes? (hand-to-sorted-string hand) %) valid_large)) 40
      :else 0
      )
    )
  )

;; TODO: a function that enumerates the keywords for the singles
(defn score-hand [hand]
  "Returns a map of all the possible scores for a hand"
  {:chance (score-chance hand)
   :ones (score-singles hand 1)
   :twos (score-singles hand 2)
   :threes (score-singles hand 3)
   :fours (score-singles hand 4)
   :fives (score-singles hand 5)
   :sixes (score-singles hand 6)
   :three-of-a-kind (score-of-a-kind hand 3)
   :four-of-a-kind (score-of-a-kind hand 4)
   :full-house (score-full-house hand)
   :small-straight (score-straight hand 4)
   :large-straight (score-straight hand 5)
   :yahtzee (score-of-a-kind hand 5)
   }
  )

(defn -main
  "Runs the game"
  [& args]
  (let [hand (roll-hand)]
    (println hand (score-hand hand))))

